packages/frontend/src/components/QrReader/index.tsx

```diff
@@ -13,360 +15,475 @@ import useTranslationFunction from '../../hooks/useTranslationFunction'
-
-export default function QrReader({ onError, onScanSuccess }: Props) {
-  const tx = useTranslationFunction()
-  const { openContextMenu } = useContext(ContextMenuContext)
-
-  const videoRef = useRef<HTMLVideoElement>(null)
-  const canvasRef = useRef<HTMLCanvasElement>(null)
-  const inputRef = useRef<HTMLInputElement>(null)
-
-  const [ready, setReady] = useState(false)
-  const [cameraAccessError, setCameraAccessError] = useState(false)
-  const [videoDevices, setVideoDevices] = useState<MediaDeviceInfo[]>([])
-  const [deviceId, setDeviceId] = useState<string | undefined>(undefined)
-  const [dimensions, setDimensions] = useState<ImageDimensions>({
-    width: 640,
-    height: 480,
-  })
-
-  // Get all current video devices available to the user.
-  useEffect(() => {
-    const getAllCameras = async () => {
-      const devices = await navigator.mediaDevices.enumerateDevices()
-      setVideoDevices(
-        devices.filter(device => {
-          return device.kind === 'videoinput'
-        })
-      )
+let worker: Worker
 
-      // Automatically select first available video device
-      if (devices.length > 0) {
-        setDeviceId(devices[0].deviceId)
-      }
-    }
-
-    getAllCameras()
-  }, [])
-
-  // General handler for errors which might occur during scanning.
-  const handleError = useCallback(
-    (error: any) => {
-      if (typeof error === 'string') {
-        onError(error)
-      } else {
-        onError(error.toString())
+let workerClipBoard: Worker
+
+export type QrCodeScanRef = {
+  handlePasteFromClipboard: () => void
+}
+
+/**
+ * component to read and process QR codes either
+ * by constantly reading image data from a video stream
+ * provided by a camera or by processing image data from
+ * clipboard or an imported image file
+ *
+ * used in Qr dialog (QrCode) in second tab
+ */
+export const QrReader = forwardRef<QrCodeScanRef, Props>(
+  ({ onError, onScanSuccess }: Props, ref) => {
+    const tx = useTranslationFunction()
+    const { openContextMenu } = useContext(ContextMenuContext)
+
+    const videoRef = useRef<HTMLVideoElement>(null)
+    const canvasRef = useRef<OffscreenCanvas>(new OffscreenCanvas(640, 480))
+    const inputRef = useRef<HTMLInputElement>(null)
+
+    const [ready, setReady] = useState(false)
+    const [cameraAccessError, setCameraAccessError] = useState(false)
+    const [videoDevices, setVideoDevices] = useState<MediaDeviceInfo[]>([])
+    const [deviceId, setDeviceId] = useState<string | undefined>(undefined)
+    const [processingFile, setProcessingFile] = useState(false)
+
+    useEffect(() => {
+      worker = new QrWorker() as Worker
+      workerClipBoard = new QrWorker() as Worker
+      return () => {
+        // terminate worker if component is unmounted,
+        // otherwise we might get a delayed error message
+        // if scan fails, even after closing the QR dialog
+        worker.terminate()
+        workerClipBoard.terminate()
       }
```

Comment:
If we go this route, then we should have the workers scoped to the component instance, and not globally.